doc/TEMPLATES.md

```diff
@@ -20,56 +23,335 @@ static template(html, { map }) {
+target.setAttribute('foo', bar);
 
-// attribute boolean bindings add or remove the attribute
-el.setAttribute('foo', '');
+// attribute boolean bindings set the attribute to an empty string or remove
+target.setAttribute('foo', ''); // when bar is truthy
+target.removeAttribute('foo'); // when bar is falsy
+
+// attribute defined bindings set the attribute if the value is non-nullish
+target.setAttribute('foo', bar); // when bar is non-nullish
+target.removeAttribute('foo'); // when bar is nullish
 
 // property bindings assign the value to the property of the node
-el.foo = bar;
+target.foo = bar;
 
-// text bindings assign the value to the content of the node
-el.textContent = foo;
+// content bindings create text nodes for basic content
+const text = document.createTextNode('');
+text.textContent = foo;
+target.append(text);
 
-// content bindings create and append text to the node
-el.appendChild(document.createTextNode(foo))
-```
+// content bindings append a child for singular, nested content
+target.append(foo);
 
-### Important note on serialization during data binding:
-Browsers must serialize values assigned to a DOM node's attributes using `toString()`. Properties with scalar types `String`, `Number`, and `Boolean` are bound to attributes using this native serialization. To help you avoid `[object Object]` surprises, properties defined using `x-element` allow you to specify their anticipated type. Attempting to bind non-scalar types to _attributes_ will result in an `x-element` error message.
+// content binding maps and appends children for arrays of nested content
+target.append(...foo);
+```
 
-The following directives are supported:
+**Important note on serialization during data binding:**
 
-* ifDefined
-* live
-* map
-* nullish
-* repeat
-* unsafeHTML
-* unsafeSVG
+Browsers must serialize values assigned to a DOM node’s attributes using
+`toString()`. It only makes sense to bind serializable values to attributes
+(i.e., `String`, `Number`, and `Boolean`). To avoid `[object Object]` surprises,
+you can specify the anticipated type of an `x-element` property which will
+throw a runtime error if it’s ever set to a value of an unanticipated type.
 
 The following template languages are supported:
 
-* html
-* svg
+* `html`
+* `svg`
+
+The following value updaters are supported:
+
+* `map` (can be used with content bindings)
+* `unsafe` (can be used with content bindings)
+* `live` (can be used with property bindings)
+
+**A note on non-primitive data:**
+
+Because DOM manipulation is *slow* — template engines do their best to avoid it
+(if a value hasn’t changed, DOM manipulation is skipped). To know if a
+non-primitive property has changed, a change-by-reference check is performed
+(versus a change-by-value check). Therefore, treat non-primitives as if they’re
+immutable. Do this: `element.property = { ...element.property, foo: 'bar' }`,
+and don’t do this: `element.property.foo = 'bar'`.
+
+## Value binding (in detail)
```

Comment:
As discussed, I tried to clearly document all the different binding-value pairs that were sensible. We may need to shift some of this stuff around in the future, but I think this starting to be a pretty complete reference.